lib/judge.rb

```diff
@@ -12,18 +12,32 @@ def self.create_task(submit)
   end
 
   def self.start_receive_results
-    rabbit = RabbitMQ.new(Rails.configuration.tasks_results_queue_name, Rails.configuration.rabbitmq_uri)
-    rabbit.start_receiving do |payload|
-      result = TaskResult.decode(payload)
-
-      Submit.transaction do
-        submit_status = create_tests_results(result)
-        Submit.find(result.result_id).update(status: submit_status)
-      end
+    rabbit = RabbitMQ.new(
+      Rails.configuration.tasks_results_queue_name,
+      Rails.configuration.rabbitmq_uri,
+      Rails.configuration.tasks_results_error_queue_name
+    )
+    rabbit.start_receiving do |payload, delivery_info|
+      Judge.handle_task_result(rabbit, payload, delivery_info)
     end
   end
 
   private_class_method
+  def self.handle_task_result(rabbit, payload, delivery_info)
+    result = TaskResult.decode(payload)
+
+    Submit.transaction do
+      submit_status = create_tests_results(result)
+      Submit.find(result.result_id).update(status: submit_status)
+      rabbit.ack(delivery_info.delivery_tag)
+    end
+  rescue ActiveRecord::RecordInvalid
```

Comment:
a było to wewnątrz bloku `transaction` bo nie kojarzę żebyśmy mieli go gdziekolwiek indziej poza tym miejscem? Ale jak najbardziej warto sprawdzić ponieważ stosunkowo ciężko to informacje w tym temacie.